conversations/impl/src/main/java/org/sakaiproject/conversations/impl/repository/TopicStatusRepositoryImpl.java

```diff
@@ -49,6 +54,66 @@ public Optional<TopicStatus> findByTopicIdAndUserId(String topicId, String userI
 
         return session.createQuery(query).uniqueResultOptional();
     }
+    
+    @Transactional
+    public TopicStatus saveTopicStatus(String topicId, String userId, boolean viewed) {
+        Session session = sessionFactory.getCurrentSession();
+        
+        try {
+            // Try to find the existing record first
+            CriteriaBuilder cb = session.getCriteriaBuilder();
+            CriteriaQuery<TopicStatus> query = cb.createQuery(TopicStatus.class);
+            Root<TopicStatus> root = query.from(TopicStatus.class);
+            query.where(
+                cb.and(
+                    cb.equal(root.get("topic").get("id"), topicId),
+                    cb.equal(root.get("userId"), userId)
+                )
+            );
+            
+            TopicStatus status = session.createQuery(query).uniqueResultOptional().orElse(null);
+            
+            if (status != null) {
+                // Update existing record
+                status.setViewed(viewed);
+                session.merge(status);
+                return status;
+            } else {
+                // Get the topic entity
+                CriteriaQuery<ConversationsTopic> topicQuery = cb.createQuery(ConversationsTopic.class);
+                Root<ConversationsTopic> topicRoot = topicQuery.from(ConversationsTopic.class);
+                topicQuery.where(cb.equal(topicRoot.get("id"), topicId));
+                ConversationsTopic topic = session.createQuery(topicQuery).uniqueResult();
+                
+                if (topic == null) {
+                    throw new IllegalArgumentException("No topic found for id: " + topicId);
+                }
+                
+                // Create and save new status
+                TopicStatus newStatus = new TopicStatus(topic, userId);
+                newStatus.setViewed(viewed);
+                try {
+                    session.save(newStatus);
+                    return newStatus;
+                } catch (ConstraintViolationException | DataIntegrityViolationException e) {
+                    // Another thread created the record first, so let's get it and update it
+                    log.debug("Caught concurrent creation exception, fetching the existing record: {}", e.getMessage());
+                    session.clear(); // Clear session to avoid stale data
+                    
+                    // Fetch the record that was just created by another thread
+                    status = session.createQuery(query).uniqueResultOptional()
+                        .orElseThrow(() -> new IllegalStateException("Expected to find TopicStatus after constraint violation"));
+                    
+                    status.setViewed(viewed);
+                    session.merge(status);
+                    return status;
```

Comment:
return the managed entity
```suggestion
                    return session.merge(status);
```

even though in this case it may be fine because the entity is loaded in this block, meaning that *status* and the entity returned by calling merge are the same Object.